presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1282,6 +1292,46 @@ private String getMaterializedViewSQL(
             return SqlFormatterUtil.getFormattedSql(unionQuery, sqlParser, Optional.empty());
         }
 
+        private Optional<Query> optimizeQueryUsingMaterializedView(Query node)
+        {
+            Map<String, List<String>> baseTableToMaterializedViewMapping = getBaseToMaterializedViewMapping();
+            MaterializedViewValidationForQueryRewrite rewriteValidator = new MaterializedViewValidationForQueryRewrite<>(baseTableToMaterializedViewMapping);
+            rewriteValidator.process(node);
+            Set<String> materializedViewCandidate = rewriteValidator.generateMaterializedViewCandidate();
+            if (materializedViewCandidate.isEmpty()) {
+                return Optional.empty();
+            }
+            analysis.setStatementRewrite();
+            Query rewriteQuery = getQueryWithMaterializedViewOptimization(node, materializedViewCandidate);
+            return Optional.of(rewriteQuery);
+        }
+
+        // TODO: The mapping should be fetched from metastore
+        private Map<String, List<String>> getBaseToMaterializedViewMapping()
+        {
+            Map<String, List<String>> baseTableToMaterializedViewMap = new HashMap<>();
+            return baseTableToMaterializedViewMap;
+        }
+
+        private Query getQueryWithMaterializedViewOptimization(
+                Query statement,
+                Set<String> materializedViewList)
```

Comment:
I don't we need a list here. Only one MV candidate for optimization is good. As mentioned above, our visitor would return only one candidate.